File: programming-skills-studyplan/Go/design_underground_system.go
package main

/*
An underground railway system is keeping track of customer travel times between different stations.
They are using this data to calculate the average time it takes to travel from one station to another.
https://leetcode.com/problems/design-underground-system/description/?envType=study-plan-v2&envId=programming-skills
*/

type UndergroundSystem struct {
	Journey map[int]CheckInData
	Time    map[string]StationAToB
}

type StationAToB struct {
	TotalTime    float64
	TotalJourney float64
}

type CheckInData struct {
	CheckIn string
	InTime  int
}

func UndergroundSystemConstructor() UndergroundSystem {
	Journey := make(map[int]CheckInData)
	TotalTime := make(map[string]StationAToB)
	return UndergroundSystem{
		Journey: Journey,
		Time:    TotalTime,
	}
}

func (this *UndergroundSystem) CheckIn(id int, stationName string, t int) {
	this.Journey[id] = CheckInData{CheckIn: stationName, InTime: t}
}

func (this *UndergroundSystem) CheckOut(id int, stationName string, t int) {
	if _, exists := this.Journey[id]; !exists {
		return
	}
	stationAToB := this.Journey[id].CheckIn + ":" + stationName
	if info, exists := this.Time[stationAToB]; exists {
		info.TotalTime += float64(t - this.Journey[id].InTime)
		info.TotalJourney++
		this.Time[stationAToB] = info
	} else {
		this.Time[stationAToB] = StationAToB{
			TotalTime:    float64(t - this.Journey[id].InTime),
			TotalJourney: 1.0,
		}
	}
	delete(this.Journey, id)

}

func (this *UndergroundSystem) GetAverageTime(startStation string, endStation string) float64 {
	AtoB := startStation + ":" + endStation
	if _, exists := this.Time[AtoB]; !exists || startStation == "" || endStation == "" {
		return 0.0
	}
	return (this.Time[AtoB].TotalTime / this.Time[AtoB].TotalJourney)
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * obj := Constructor();
 * obj.CheckIn(id,stationName,t);
 * obj.CheckOut(id,stationName,t);
 * param_3 := obj.GetAverageTime(startStation,endStation);
 */

File: programming-skills-studyplan/Java/designUnderGroundSystem.java
/*
An underground railway system is keeping track of customer travel times between different stations.
They are using this data to calculate the average time it takes to travel from one station to another.
https://leetcode.com/problems/design-underground-system/description/?envType=study-plan-v2&envId=programming-skills
*/
class TravelInfo {
    public double totalTime,totalTravel;
    public TravelInfo(double totalTime, double totalTravel) {
        this.totalTime = totalTime;
        this.totalTravel = totalTravel;
    }
}

class Travel {
    public double inTime;
    public String station;
    public Travel(double inTime,String station) {
        this.inTime = inTime;
        this.station = station;
    }
}

class UndergroundSystem {
    private HashMap<Integer,Travel> checkInInfo;
    private HashMap<String,TravelInfo> totalInfo;

    public UndergroundSystem() {
        this.checkInInfo = new HashMap<>();
        this.totalInfo = new HashMap<>();
    }
    
    public void checkIn(int id, String stationName, int t) {
        this.checkInInfo.put(id, new Travel(t,stationName));
    }
    
    public void checkOut(int id, String stationName, int t) {
        if (!this.checkInInfo.containsKey(id)) {
            return;
        }
        String key = this.checkInInfo.get(id).station + ":" + stationName;
        if (this.totalInfo.containsKey(key)) {
            TravelInfo val = this.totalInfo.get(key);
            val.totalTime += (t - this.checkInInfo.get(id).inTime);
            val.totalTravel++;
            this.totalInfo.put(key,val);
        } else {
            this.totalInfo.put(key,new TravelInfo((t - this.checkInInfo.get(id).inTime) ,1.0));
        }
        this.checkInInfo.remove(id);
    }
    
    public double getAverageTime(String startStation, String endStation) {
        String key = startStation + ":" + endStation;
        if (!this.totalInfo.containsKey(key) || this.totalInfo.get(key).totalTravel == 0 ) {
            return 0.0;
        }
        TravelInfo val = this.totalInfo.get(key);
        return val.totalTime / val.totalTravel;
    }
}

/**
 * Your UndergroundSystem object will be instantiated and called as such:
 * UndergroundSystem obj = new UndergroundSystem();
 * obj.checkIn(id,stationName,t);
 * obj.checkOut(id,stationName,t);
 * double param_3 = obj.getAverageTime(startStation,endStation);
 */